File: Ejercicios/T4/CODIGO/PYTHON/ContornosGradientes/sobel.py
import cv2
import numpy as np

filename = "../../data/Contour.png"
image = cv2.imread(filename, cv2.IMREAD_GRAYSCALE)
if image is None:
    print("Non poiden ler a imaxe")



# Sobel na direccion x
sobelx = cv2.Sobel(image, cv2.CV_32F, 1, 0)
# Sobel na direccion y
sobely = cv2.Sobel(image,cv2.CV_32F,0,1)

# Normalizamos para visualizacion
cv2.normalize(sobelx,
                dst = sobelx,
                alpha = 0,
                beta = 1,
                norm_type = cv2.NORM_MINMAX,
                dtype = cv2.CV_32F)
cv2.normalize(sobely,
                dst = sobely,
                alpha = 0,
                beta = 1,
                norm_type = cv2.NORM_MINMAX,
                dtype = cv2.CV_32F)
cv2.imshow("Gradiente de Sobel X", sobelx)
cv2.imshow("Gradiente de Sobel  Y", sobely)
cv2.waitKey(0)

File: Ejercicios/T4/CODIGO/C++/ContornosGradientes/sobel.cpp
#include <iostream>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

using namespace std;
using namespace cv;


int main(int argc, char** argv){

  //Lemos os argumentos de entrada
  const char* file = argv[1];
  Mat image = imread(file, IMREAD_GRAYSCALE);

  if(image.empty())
  {
    cout << "Erro lendo a imaxe" << file<< endl;
    return -1;
  }

Mat sobelx, sobely;

// Aplicamos o Sobel na direccion x
Sobel(image, sobelx, CV_32F, 1, 0);

// Aplicamos o Sobel na direccion y
Sobel(image, sobely, CV_32F, 0, 1);

// Normalizamos a imaxe para visualizacion
normalize(sobelx, sobelx, 0, 1, NORM_MINMAX);
normalize(sobely, sobely, 0, 1, NORM_MINMAX);

imshow("Gradientes X", sobelx);
waitKey(0);
imshow("Gradientes Y", sobely);
waitKey(0);

return 0;
}
